File: app/src/main/java/com/elewise/ldmobile/ui/DocsFragment.java
package com.elewise.ldmobile.ui;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.elewise.ldmobile.R;
import com.elewise.ldmobile.model.*;
import com.elewise.ldmobile.api.*;
import com.elewise.ldmobile.service.Session;
import com.elewise.ldmobile.utils.ImageUtils;
import com.elewise.ldmobile.utils.MessageUtils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class DocsFragment extends Fragment {
    private TextView tvNoResults;
    private ListView lvDocs;
    public static final String ARG_PAGE = "ARG_PAGE";
    private ProcessType processType;
    private ProgressDialog progressDialog;
    AlertDialog dialog;

    public static DocsFragment newInstance(ProcessType processType) {
        Bundle args = new Bundle();
        args.putString(ARG_PAGE, processType.name());
        DocsFragment fragment = new DocsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            processType = ProcessType.valueOf(getArguments().getString(ARG_PAGE));
        }
        progressDialog = new ProgressDialog(getActivity());
        progressDialog.setCancelable(false);
        progressDialog.setMessage(getString(R.string.progress_dialog_load));
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                                       Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_docs, container, false);
        tvNoResults = view.findViewById(R.id.tvNoResults);
        lvDocs = view.findViewById(R.id.lvDocs);
        registerForContextMenu(lvDocs);

        // set adapter
        final DocsAdapter adapter = new DocsAdapter(getContext(), new ArrayList<>());
        lvDocs.setAdapter(adapter);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        getDocuments();
    }

    private void getDocuments() {
        progressDialog.show();
        new Thread(() -> {
            FilterData[] filterData = Session.getInstance().getFilterData();
            List<DocumentForList> docsList = Session.getInstance().getDocuments(10, 0,
                    processType, filterData);

            handleDocumentsResponse(docsList);
        }).start();
    }

    private void handleDocumentsResponse(List<DocumentForList> docsList) {
        getActivity().runOnUiThread(() -> {
            progressDialog.cancel();

            if (docsList != null && !docsList.isEmpty()) {
                ((DocsAdapter) lvDocs.getAdapter()).setList(docsList);
                lvDocs.setVisibility(View.VISIBLE);
                tvNoResults.setVisibility(View.GONE);
            } else {
                lvDocs.setVisibility(View.GONE);
                tvNoResults.setVisibility(View.VISIBLE);
            }
        });
    }

    private void showDocDetail(final Document document) {
        progressDialog.show();
        new Thread(() -> {
            try {
                ParamDocumentDetailsResponse result = Session.getInstance().getDocumentDetail(document.getDoc_id(), document.getDoc_type());
                if (result == null) {
                    showError(getString(R.string.error_unknown));
                } else {
                    if (result.getStatus().equals(ResponseStatusType.E.name())) {
                        String message = result.getMessage();
                        if (TextUtils.isEmpty(message)) showError(getString(R.string.error_unknown));
                        else showError(message);
                    } else if (result.getStatus().equals(ResponseStatusType.A.name())) {
                        showError(getString(R.string.error_authentication));
                    } else {
                        Session.getInstance().setCurrentDocumentDetail(result);
                        handleDocumentDetailsResponse(result);
                    }
                }
            } catch (Exception e) {
                showError(getString(R.string.error_unknown));
                e.printStackTrace();
            }
        }).start();
    }

    private void showError(String message) {
        getActivity().runOnUiThread(() -> {
            progressDialog.cancel();
            Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
        });
    }

    private void handleDocumentDetailsResponse(final ParamDocumentDetailsResponse documentDetail) {
        getActivity().runOnUiThread(() -> {
            progressDialog.cancel();

            if (documentDetail.getDoc_type().equals(DocType.PD.name())) {
                Intent intent = new Intent();
                intent.setClass(DocsFragment.this.getActivity(), DocPacketActivity.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                DocsFragment.this.getActivity().startActivity(intent);
            } else {
                Intent intent = new Intent();
                intent.setClass(DocsFragment.this.getActivity(), DocActivity.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                DocsFragment.this.getActivity().startActivity(intent);
            }
        });
    }


    @Override
    public void onDestroy() {
        if (progressDialog!= null) progressDialog.dismiss();
        if (dialog != null) dialog.dismiss();

        super.onDestroy();
    }

    /**
     * Adapter
     */
    public class DocsAdapter extends BaseAdapter {
        private Context context;
        private List<DocumentForList> list;
        private SimpleDateFormat sdf = new SimpleDateFormat("dd.MM.yyyy");

        public DocsAdapter(Context context, List<DocumentForList> item) {
            this.context = context;
            this.list = item;
        }

        public void setList(List<DocumentForList> list) {
            this.list = list;
            notifyDataSetChanged();
        }

        @Override
        public int getCount() {
            return list.size();
        }

        @Override
        public Object getItem(int position) {
            return list.get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            if (list.get(position).isSection()) {
                // if section header
                convertView = inflater.inflate(R.layout.list_group_date, parent, false);
                TextView tvSectionTitle = convertView.findViewById(R.id.tvSectionTitle);
                tvSectionTitle.setText(String.format(getString(R.string.docs_activity_date_group), list.get(position).getSectionTitle()));
                if (sdf.format(new Date()).equals(list.get(position).getSectionTitle())) {
                    tvSectionTitle.setTypeface(null, Typeface.BOLD);
                    tvSectionTitle.setTextColor(getResources().getColor(R.color.colorPrimary));
                } else {
                    tvSectionTitle.setTypeface(null, Typeface.NORMAL);
                    tvSectionTitle.setTextColor(getResources().getColor(R.color.colorAccent));
                }
            } else {
                // if list
                convertView = inflater.inflate(R.layout.list_doc_item, parent, false);
                TextView tvDocTitle = convertView.findViewById(R.id.tvDocTitle);
                TextView tvDocBody = convertView.findViewById(R.id.tvDocBody);
                ImageView imgDocType = convertView.findViewById(R.id.imgDocType);
                ImageView imgAttache = convertView.findViewById(R.id.imgAttache);
                ImageView imgAction = convertView.findViewById(R.id.imgAction);

                final Document document = list.get(position).getDocument();
                tvDocTitle.setText(document.getContractor());
                tvDocBody.setText(document.getDoc_name());
                if (document.getAttach_flag()) {
                    imgAttache.setVisibility(View.VISIBLE);
                } else {
                    imgAttache.setVisibility(View.INVISIBLE);
                }

                ImageUtils.INSTANCE.setIcon(getResources(), imgDocType, document.getDoc_icon());

                if (document.getDoc_type().equals(DocType.PD.name())) {
                    imgAction.setVisibility(View.GONE);
                } else {
                    imgAction.setVisibility(View.VISIBLE);
                    ImageUtils.INSTANCE.setIcon(getResources(), imgAction, document.getAction_icon());
                }

                convertView.setOnClickListener(view -> showDocDetail(document));
            }

            return convertView;
        }
    }
}

File: app/src/main/java/com/elewise/ldmobile/ui/DocsFragment.kt
package com.elewise.ldmobile.ui

import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.elewise.ldmobile.R
import com.elewise.ldmobile.api.*
import com.elewise.ldmobile.model.DocType
import com.elewise.ldmobile.model.DocumentForList
import com.elewise.ldmobile.model.ProcessType
import com.elewise.ldmobile.service.Session
import com.elewise.ldmobile.utils.ImageUtils.setIcon
import com.elewise.ldmobile.utils.MessageUtils
import kotlinx.android.synthetic.main.fragment_docs.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*


class DocsFragment : Fragment() {

    private var viewModelJob = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + viewModelJob)

    private var processType: ProcessType? = null
    private var progressDialog: ProgressDialog? = null
    private var dialog: AlertDialog? = null
    private var session = Session.getInstance()
    private var currrentDocFrom = 0
    private lateinit var adapter: DocsAdapter
    private var lastFilterData: List<FilterData> = listOf()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            processType = ProcessType.valueOf(it.getString(ARG_PAGE)!!)
        }
    }

    private fun showProgressDialog() {
        if (progressDialog == null) {
            progressDialog = ProgressDialog(activity).apply {
                setCancelable(false)
                setMessage(getString(R.string.progress_dialog_load))
            }
        }
        progressDialog?.show()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_docs, container, false)
        adapter = DocsAdapter(context!!, ArrayList())
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
//        registerForContextMenu(lvDocs)

        lvDocs.setOnScrollListener(object : AbsListView.OnScrollListener {
            override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {
                if (scrollState == AbsListView.OnScrollListener.SCROLL_STATE_IDLE) {
                    if (lvDocs.getLastVisiblePosition() - lvDocs.getHeaderViewsCount() -
                        lvDocs.getFooterViewsCount() >= lvDocs.adapter.getCount() - 1) {
                        getDocuments(TYPE_LOAD_DATA.INCR)
                    }

                    if (lvDocs.firstVisiblePosition == 0) {
                        getDocuments(TYPE_LOAD_DATA.DECR)
                    }
                }
            }

            override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {}
        })

        llProgressBar.setOnClickListener { /* this need empty */}

        // set adapter
        lvDocs.setAdapter(adapter)
    }

    override fun onStart() {
        super.onStart()
        getDocuments()
    }

    private enum class TYPE_LOAD_DATA {
        NONE, INCR, DECR
    }

    private fun getDocuments(typeLoadData: TYPE_LOAD_DATA = TYPE_LOAD_DATA.NONE) {
        uiScope.launch {
            try {
                showProgressBar(true)

                if (session.filterData.isEmpty()) {
                    loadFilterFromServer()
                }

                // определяем параметры загрузки
                val tempDocsList = arrayListOf<DocumentForList>()
                val from = when (typeLoadData) {
                    TYPE_LOAD_DATA.INCR -> {
                        tempDocsList.addAll(adapter.list)
                        currrentDocFrom += session.docSize
                        currrentDocFrom
                    }
                    TYPE_LOAD_DATA.DECR -> {
                        currrentDocFrom = 0
                        currrentDocFrom
                    }
                    else -> {
                        if (lastFilterData == session.filterData) {
                            // фильтр не изменился. не выполняем загрузку
                            showProgressBar(false)
                            return@launch
                        } else {
                            adapter.setItemsList(listOf())
                            0
                        }
                    }
                }

                // обновим сохраненное в форме значение текущего фильтра
                lastFilterData = session.filterData

                // запросим документы
                val request = session.getDocuments(from, processType, lastFilterData)
                val response = request.await().body()

                if (response != null) {
                    if (response.status == ResponseStatusType.S.name) {
                        tempDocsList.addAll(session.groupDocByDate(response.group_flag, response.contents))
                        if (tempDocsList.isNotEmpty()) {
                            adapter.setItemsList(tempDocsList)
                            if (typeLoadData == TYPE_LOAD_DATA.INCR && adapter.list.size - 1 > lvDocs.lastVisiblePosition) {
                                // скролл на одну позицию вниз с анимацией
                                lvDocs.smoothScrollToPosition(lvDocs.lastVisiblePosition + 1)
                            }
                            lvDocs.visibility = View.VISIBLE
                            tvNoResults.visibility = View.GONE
                        } else {
                            lvDocs.visibility = View.GONE
                            tvNoResults.visibility = View.VISIBLE
                        }
                    } else if (response.status == ResponseStatusType.E.name) {
                        var errorMessage = getString(R.string.error_load_data)
                        response.message?.let {
                            if (it.isNotEmpty()) errorMessage = it
                        }
                        showError(errorMessage)
                    } else if (response.status == ResponseStatusType.A.name) {
                        session.errorAuth()
                    } else {
                        showError(getString(R.string.error_unknown_status_type))
                    }
                } else {
                    showError()
                }
            } catch (e: Exception) {
                showError()
            }

            showProgressBar(false)
        }
    }

    private fun showProgressBar(isVisible: Boolean) {
        if (isVisible) {
            llProgressBar.visibility = View.VISIBLE
        } else {
            llProgressBar.visibility = View.GONE
        }
    }

    @Throws(Exception::class)
    private suspend fun loadFilterFromServer() {
        // считаем фильтр с сервера
        val filterData = arrayListOf<FilterData>()
        session.filterSettings.await().body()?.let { response ->
            response.filters?.forEach {
                if (it.last_value?.isNotEmpty() == true || it.last_value2?.isNotEmpty() == true) {
                    filterData.add(FilterData(it.name, it.last_value, it.last_value2))
                }
            }
        }
        session.filterData = filterData
    }

    private fun showDocDetail(document: Document) {
        uiScope.launch {
            try {
                showProgressDialog()
                val request = Session.getInstance().getDocumentDetail(document.doc_id)
                val response = request.await()
                handleDocumentDetailsResponse(response.body())
            } catch (e: Exception) {
                progressDialog?.dismiss()
                showError()
            }
        }
    }

    private fun handleDocumentDetailsResponse(response: ParamDocumentDetailsResponse?) {
        progressDialog?.dismiss()
        var errorMessage: String = getString(R.string.error_load_data)
        if (response != null) {
            if (response.status == ResponseStatusType.S.name) {
                Session.getInstance().currentDocumentDetail = response
                val intent = Intent()
                intent.setClass(activity!!, DocActivity::class.java)
                startActivity(intent)
                return
            } else if (response.status == ResponseStatusType.E.name) {
                response.message?.let { if (it.isNotEmpty()) errorMessage = it }
            } else if (response.status == ResponseStatusType.A.name) {
                session.errorAuth()
                return
            } else {
                errorMessage = getString(R.string.error_unknown_status_type)
            }
        }
        showError(errorMessage)
    }

    private fun showError(errorMessage: String = getString(R.string.error_load_data)) {
        // показать ошибку
        dialog = MessageUtils.createDialog(activity, getString(R.string.alert_dialog_error), errorMessage).apply { show() }
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModelJob.cancel()
        progressDialog?.dismiss()
        dialog?.dismiss()
    }

    /**
     * Adapter
     */
    inner class DocsAdapter(private val context: Context, var list: List<DocumentForList>) : BaseAdapter() {
        private val sdf = SimpleDateFormat("dd.MM.yyyy")

        fun setItemsList(list: List<DocumentForList>) {
            this.list = list
            notifyDataSetChanged()
        }

        override fun getCount(): Int {
            return if (list.isNotEmpty()) {
                list.size
            } else 0
        }

        override fun getItem(position: Int): Any {
            return list[position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view: View
            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            if (list[position].isSection) {
                // if section header
                view = inflater.inflate(R.layout.list_group_date, parent, false)
                val tvSectionTitle = view.findViewById<TextView>(R.id.tvSectionTitle)
                tvSectionTitle.text = list[position].sectionTitle
                if (sdf.format(Date()) == list[position].sectionTitle) {
                    tvSectionTitle.setTypeface(null, Typeface.BOLD)
                    tvSectionTitle.setTextColor(resources.getColor(R.color.colorPrimary))
                } else {
                    tvSectionTitle.setTypeface(null, Typeface.NORMAL)
                    tvSectionTitle.setTextColor(resources.getColor(R.color.colorAccent))
                }
            } else {
                // if list
                view = inflater.inflate(R.layout.list_doc_item, parent, false)
                val tvDocTitle = view.findViewById<TextView>(R.id.tvDocTitle)
                val tvDocBody = view.findViewById<TextView>(R.id.tvDocBody)
                val imgDocType = view.findViewById<ImageView>(R.id.imgDocType)
                val imgAttache = view.findViewById<ImageView>(R.id.imgAttache)
                val imgAction = view.findViewById<ImageView>(R.id.imgAction)
                val document = list[position].document
                tvDocTitle.text = document.contractor
                tvDocBody.text = document.doc_name
                if (document.attach_flag == true) {
                    imgAttache.visibility = View.VISIBLE
                } else {
                    imgAttache.visibility = View.INVISIBLE
                }
                setIcon(resources, imgDocType, document.doc_icon)
                imgAction.visibility = View.VISIBLE
                setIcon(resources, imgAction, document.action_icon)
                view.setOnClickListener { v: View? -> showDocDetail(document) }
            }
            return view
        }
    }

    companion object {
        const val ARG_PAGE = "ARG_PAGE"
        @JvmStatic
        fun newInstance(processType: ProcessType): DocsFragment {
            val args = Bundle()
            args.putString(ARG_PAGE, processType.name)
            val fragment = DocsFragment()
            fragment.arguments = args
            return fragment
        }
    }
}